Vulkan AS rebuild-on-replay: Copy AS input buffers on build

cmannett85-arm opened the pull request:
When `vkCmdBuildAccelerationStructuresKHR` is called, the input buffers used for the build are copied into host-accessible memory, so when initial state serialisation is triggered it can be copied into the capture file.

To unify the behaviour of the different geometry types, there is only one readback buffer per AS, and in the case of triangle data it can contain the vertex, index, and transform data concatenated into a single contiguous block.

The buffer data along with build arguments are stored in a `VkAccelerationStructureInfo` object owned by the AS's `VkResourceRecord`.  If this record is freed, then the copied buffer mem is also freed.  If this destruction occurs during the active capture phase, the AS info is kept alive until the end of the active capture.

Serialisation of the initial state will follow in later patches.

Caveats:
* `vkCmdBuildAccelerationStructuresIndirectKHR` is not handled
* `VkAccelerationStructureGeometryInstancesDataKHR::arrayOfPointers` mode is not handled

baldurk commented on renderdoc/driver/vulkan/wrappers/vk_misc_funcs.cpp:
```diff
@@ -335,6 +334,36 @@ void WrappedVulkan::vkDestroyImage(VkDevice device, VkImage obj, const VkAllocat
+    return;
+
+  // artificially extend the lifespan of the AS info to ensure it isn't destroyed before
+  // serialisation completes
```
I don't think this should be necessary because the capture process already refcounts the resource record to keep it alive for the duration of capture. Otherwise we would lose the chunks contained within. Anything that's marked resource referenced also adds a reference on the resource record. This doesn't protect against applications deleting resources referenced by a command buffer, but that would be illegal in vulkan (if the command buffer is later submitted).

Did you see this happening in practice or was this a theoretical fix?

cmannett85-arm replied:
I did see it happening but given your explanation, perhaps I was misinterpreting.  I'll remove this and give your AS info refcounting idea a go, and that should solve the fiddly lifetime issues I was tackling.

baldurk commented on renderdoc/driver/vulkan/vk_acceleration_structure.h:
```diff
@@ -28,9 +28,52 @@
+    struct Instances
+    {
+      // arrayOfPointers TODO
+    };
```
I don't think we should have the BLASs referenced by a TLAS in the same way as geometry. This would be better stored as a `rdcarray<VkAccelerationStructureKHR> asData;` next to `geometryData` to more clearly indicate that only one is present.

cmannett85-arm replied:
I don't follow what you're saying, this struct is meant to represent the non-memory state of `VkAccelerationStructureGeometryInstancesDataKHR` but I'm not handling `arrayOfPointers` hence why it is empty.  The actual `VkAccelerationStructureInstanceKHR` data is copied into `readbackMem` like everything else.

baldurk replied:
Ah sorry I misunderstood what this is for.

With `arrayOfPointers` what we want to store isn't different we just can't do it in one copy. We'll probably need to do an inefficient compute-based copy to follow the indirections, but in that case this struct still wouldn't be needed as the data would still end up in `readbackMem` just via a more circuitous route?

cmannett85-arm replied:
Ah yes, you're absolutely right as the copy would 'flatten' the indirection so on replay `arrayOfPointers` would always be false - I'll remove it.

cmannett85-arm replied:
I'm going to need some hand holding when it comes to using a compute shader for the indirect copying - is there anything similar in the codebase already I can base it on?

baldurk replied:
I don't think there's anything similar on vulkan unfortunately. This is something uniquely difficult on RT needing this kind of GPU work at capture time.

Roughly it would mean tracking enough compute state in the command buffer during capture to push + pop around our own dispatch, and then write a shader to indirect through the pointers and copy out a BLAS-instance sized bit of memory to the output. I think since you will inevitably disturb the pipeline layout and push constants etc you can push-constant the number of BLASs.

cmannett85-arm replied:
For this patch would it be acceptable to just fail the capture if `arrayOfPointers` is set and then return to the issue after the general case is merged?  I'm worried about getting bogged down in this when I haven't even seen the mode used yet.

baldurk replied:
Oh yes sorry I didn't mean to imply you needed to add support here. As long as there's an assert that the case isn't handled it can be left as a TODO for a future PR. I only made the original comment because I misunderstood what you were needing this struct for.

baldurk commented on renderdoc/driver/vulkan/vk_acceleration_structure.h:
```diff
@@ -28,9 +28,52 @@
+  VkBuildAccelerationStructureFlagsKHR flags = 0;
+
+  rdcarray<GeometryData> geometryData;
+  rdcarray<VkAccelerationStructureBuildRangeInfoKHR> buildRangeInfo;
```
Rather than having two parallel arrays here I would prefer to have this information stored directly in `GeometryData` with one array.

I'm guessing maybe this is done so this separate build range array can be re-used as-is somewhere on build? but we still need to recreate the `VkAccelerationStructureBuildGeometryInfoKHR` array and the cost of reconstructing the build range array as a temporary will be mostly trivial in comparison - I think the ease of reading the code and seeing the related information together is worthwhile.

baldurk commented on renderdoc/driver/vulkan/vk_acceleration_structure.cpp:
```diff
@@ -360,6 +757,136 @@ void VulkanAccelerationStructureManager::Apply(ResourceId id, const VkInitialCon
+                                                            CaptureState state)
+{
+  // If we're capturing this frame, then the info could be attached to an initial state so we need
+  // to delete it later
```
At the risk of sounding like a broken record, I think refcounting on the acceleration structure info would work better here for this kind of versioning/lifetime would work better here. I think it will also come in handy further down the line when we will likely want to start shuffling this data off to disk early and so the lifetime won't be as binary.

This also has the benefit of removing the need for the deferred cleanup process entirely I think.

You would need to lock the cap transition lock here, otherwise there's a potential significant race here, but generally I try to avoid checking the active/background capturing state where possible as it can lead to hard-to-test edge cases. At least when there's another fairly simple alternative.

baldurk commented on renderdoc/driver/vulkan/vk_acceleration_structure.cpp:
```diff
@@ -37,13 +37,410 @@ constexpr VkDeviceSize handleCountSize = 8;
+          vertexData.alignment = mrq.alignment;
+          vertexData.size = vertexData.buf != VK_NULL_HANDLE
+                                ? vertexData.rao.record->memSize - vertexData.rao.offset
+                                : rangeInfo.primitiveCount * 3 * triInfo.vertexStride;
```
I think this check is incorrect - `vertexData.buf` should never be NULL here - I think this meant to check if index data is present.

We have the information from `VkAccelerationStructureGeometryTrianglesDataKHR::maxVertex` for bounding the vertex buffer size in the case where index buffers are used, so we should use it otherwise we could copy a huge amount of extra data. We could `RDCMIN` these if we're concerned about people putting in a stupid value (which now that I think about it, someone definitely will).

Also for this code I think it's better to wrap both vertex + index parts in `if(indexData)` to make the code cleaner, even if it means a little duplication.

baldurk commented on renderdoc/driver/vulkan/vk_acceleration_structure.cpp:
```diff
@@ -37,13 +37,410 @@ constexpr VkDeviceSize handleCountSize = 8;
+        const VkDeviceSize maxAlignment =
+            RDCMAX(RDCMAX(vertexData.alignment, indexData.alignment), transformData.alignment);
+
+        // We want to copy the ubput buffers into one big block so sum the sizes up together
```
typo: `input buffers`.

baldurk commented on renderdoc/driver/vulkan/vk_acceleration_structure.cpp:
```diff
@@ -37,13 +37,410 @@ constexpr VkDeviceSize handleCountSize = 8;
+
+        // Insert copy commands
+        VkBufferCopy region = {
+            vertexData.rao.offset + vertexData.start,
```
I think there's a risk of confusion here in future with the terms `offset` and `start` being somewhat interchangeable. Here `offset` meaning "offset from the beginning of the actual buffer to the address passed in" and `start` being "the relative location where the geometry data sits".

Since you are using the local `BufferData` you could maybe rename these variables and add a helper function for this, but I wonder if you need to keep `start` separate? It looks like everywhere you use this it's always added to offset since once we've processed all the offsets and sizes here we really only care about the absolute offset from buffer start to where the data sits, the address the application passed in is mostly irrelevant.

baldurk commented on renderdoc/driver/vulkan/vk_acceleration_structure.cpp:
```diff
@@ -37,13 +37,410 @@ constexpr VkDeviceSize handleCountSize = 8;
+        buildData.transformOffset = !transformData
+                                        ? 0
+                                        : (uint32_t)(AlignUp(vertexData.size, vertexData.alignment) +
+                                                     AlignUp(indexData.size, indexData.alignment));
```
I think using `ifs` here would be easier to read than ternaries.

baldurk commented on renderdoc/driver/vulkan/vk_acceleration_structure.cpp:
```diff
@@ -37,13 +37,410 @@ constexpr VkDeviceSize handleCountSize = 8;
+                                          VK_PIPELINE_STAGE_TOP_OF_PIPE_BIT, 0, 0, VK_NULL_HANDLE,
+                                          static_cast<uint32_t>(barriers.size()), barriers.data(),
+                                          0, VK_NULL_HANDLE);
+    }
```
I think rather than doing individual buffer barriers it would be easier to just do one global memory barrier. I'm not sure the exact difference for drivers but I believe few even use buffer barriers as anything specific, and if any do as soon as we have more than a couple of buffers here it's likely going to decay.

Also note that this barrier after the copies is to ensure that the input buffers aren't overwritten by any _later_ commands, so we should have both source and destination access masks to ensure that.

cmannett85-arm replied:
Is there any existing functionality that implements this?  Rather than me re-invent the wheel.

baldurk replied:
It could be the same code as you have here but just one barrier at the end instead of N.

baldurk reviewed (changes requested):
As a general note since I can't put this on the specific line, remember to disable the indirect build physical device feature.

It's fine not to support indirect builds as it seems few drivers do as well and I don't think anyone will use it given the feature is not supported in D3D12, but I think it should be technically possible as it looks like we get bounds on the data size on the CPU? The main challenge will be doing the copy, might have to fall back to a compute shader or something and do it inefficiently.

cmannett85-arm committed:
Vulkan AS rebuild-on-replay: Copy AS input buffers on build

When vkCmdBuildAccelerationStructuresKHR is called, the input buffers used for the build are copied into host-accessible memory, so when initial state serialisation is triggers it can be copied into the capture file.

To unify the behaviour of the different geometry types, there is only one readback buffer per AS, and in the case of triangle data it can contain the vertex, index, and transform data concatenated into a single contiguous block.

The buffer data along with build arguments are stored in a ref-counted VkAccelerationStructureInfo object held by the AS's VkResourceRecord.  Once the info object is destroyed, then the copied buffer mem is also freed.

Serialisation of the initial state will follow in later patches.

Caveats:
* vkCmdBuildAccelerationStructuresIndirectKHR is not handled
* VkAccelerationStructureGeometryInstancesDataKHR::arrayOfPointers mode is not handled

cmannett85-arm committed:
Replace queried AS compaction sizes with full size

The spec does not mandate that the compaction size is the same between capture and replay, so we effectively disable it by always returning the full AS size when the compaction size is queried.

baldurk commented on renderdoc/driver/vulkan/vk_acceleration_structure.h:
```diff
@@ -43,8 +87,35 @@ class VulkanAccelerationStructureManager
+  RDResult CopyInputBuffers(VkCommandBuffer commandBuffer,
+                            const VkAccelerationStructureBuildGeometryInfoKHR &info,
+                            const VkAccelerationStructureBuildRangeInfoKHR *buildRange,
+                            CaptureState state);
```
This state seems redundant here - we don't do this during replay, and we don't need to differentiate between background and active capturing. Doing a quick look through unless I'm missing something this value is (ultimately) not used.

Same applies below to `CopyAccelerationStructure`.

baldurk commented on renderdoc/driver/vulkan/vk_resources.cpp:
```diff
@@ -3948,6 +3948,107 @@ void ImgRefs::Split(bool splitAspects, bool splitLevels, bool splitLayers)
+
+void QueryPoolInfo::Reset(uint32_t firstQuery, uint32_t queryCount)
+{
+  m_Entries.removeIf([&](const auto &entry) {
```
Please avoid `auto` except where necessary for lambdas or extremely ugly for std iterators, you can use `Entry` here.

Also please use explicit lambda capture rather than unconditional capture to be clear about what is stored in the lambda.

baldurk commented on renderdoc/driver/vulkan/vk_resources.cpp:
```diff
@@ -3948,6 +3948,107 @@ void ImgRefs::Split(bool splitAspects, bool splitLevels, bool splitLayers)
+}
+
+void QueryPoolInfo::Replace(uint32_t firstQuery, uint32_t queryCount,
+                            const std::function<void(Entry)> &writeEntry) const
```
Is there a reason for this function to be separated out? It looks like it's only called from the other `Replace()` to apply any replacements so could be done inside that function? That should allow some simplification and ease of reading since it took me a couple of passes to understand the code flow with the functions being named the same.

cmannett85-arm replied:
This is called from both the public `Replace(..)` methods, to de-duplicate the logic.

baldurk commented on renderdoc/driver/vulkan/vk_resources.h:
```diff
@@ -2236,6 +2236,41 @@ inline FrameRefType MarkMemoryReferenced(std::unordered_map<ResourceId, MemRefs>
+  void Replace(uint32_t firstQuery, uint32_t queryCount,
+               const std::function<void(Entry)> &writeEntry) const;
+
+  rdcarray<Entry> m_Entries;
```
Thinking about this more generally, query pools are single-typed so we know at creation time whether we're either going to replace all results or none of them (when the query pool type is compacted size).

Do you have any empirical data or intuition on how large those query pools are? I would assume either 1, some small fixed count (UE5 seems to use 64, configurable), or maybe the application's expected max number of BLASs (some thousands?).

Unless that seems unreasonable or you're otherwise worried about having a huge number of possible queries with only a few actually being valid, I think this code could be simplified a good deal by a) only allocating this struct when needed for patching and b) pre-allocating all the entries up front. Then you don't have to worry about intersecting, resizing this array on the fly, etc. We just read all our values out of it.

cmannett85-arm replied:
That's a good point, the largest I've seen is 64 but that was for UE5 demos.  I can't find anything in the spec regarding an upper bound so I've assumed the worst.  I've asked on the Vulkan Discord if anyone knows if the spec does actually define it.

If we make the reasonable assumption that it will be a few thousand max, then allocating the buffer (as described below) using whatever size is passed into `VkQueryPoolCreateInfo::queryCount` will definitely simplify the code.

baldurk commented on renderdoc/driver/vulkan/wrappers/vk_cmd_funcs.cpp:
```diff
@@ -8009,27 +8049,35 @@ void WrappedVulkan::vkCmdWriteAccelerationStructuresPropertiesKHR(
+    for(uint32_t i = 0; i < accelerationStructureCount; ++i)
+      sizes.push_back(GetRecord(pAccelerationStructures[i])->memSize);
+
+    qpInfo->Add(firstQuery, std::move(sizes));
```
Unfortunately I've realised seeing this that there is an ordering/timeline constraint that I hadn't considered with this data. Because the properties write to a query and then the queries are resolved separately, these commands can be recorded and submitted out of order so we can't update a single authoritative fake query pool on the CPU timeline and update from there.

There may be a clever solution to this to resolve it on the CPU by deferring query updates and keeping multiple versions around, but I think I'd say we should solve it on the GPU to keep things simpler:

1. Create an internal buffer backing the query pool (ie. what the driver probably does on most GPUs) of equal size to its full set of results. It should probably be readback memory to be CPU-accessible because of 4 below. In theory in future we could have two buffers kept up to date in parallel, one for CPU reads and one for GPU reads, but I doubt that will be a problem.
2. This `vkCmdWriteAccelerationStructuresPropertiesKHR` writes the data via `vkUpdateBuffer` into that backing buffer - again effectively in parallel to what the driver is doing.
3. In `vkCmdCopyQueryPoolResults` we do a plain `vkCmdCopyBuffer` from backing buffer to user's results buffer.
4. In `vkGetQueryPoolResults` we CPU-read from the backing buffer.

It's a little bit uglier/more work than tracking purely on the CPU side but it keeps the timeline neatly in sync without having to mirror it with submission tracking and CPU-side versioning of data.

baldurk commented on renderdoc/driver/vulkan/wrappers/vk_misc_funcs.cpp:
```diff
@@ -1699,6 +1705,8 @@ void WrappedVulkan::vkResetQueryPool(VkDevice device, VkQueryPool queryPool, uin
   SERIALISE_TIME_CALL(
       ObjDisp(device)->ResetQueryPool(Unwrap(device), Unwrap(queryPool), firstQuery, queryCount));
 
+  GetRecord(queryPool)->queryPoolInfo->Reset(firstQuery, queryCount);
```
This also misses handling of `CmdResetQueryPool` but I think with the other comments handling reset may not be needed at all.

baldurk commented on renderdoc/driver/vulkan/vk_acceleration_structure.cpp:
```diff
@@ -37,13 +38,418 @@ constexpr VkDeviceSize handleCountSize = 8;
+
+        if(instanceInfo.arrayOfPointers)
+          return RDResult(ResultCode::InternalError,
+                          "AS instance build arrayOfPointers unsupported");
```
FYI there is a helper `RETURN_ERROR_RESULT` which can do this and also print the `RDCERR` for you, without the calling code needing to do that.

baldurk reviewed (changes requested):
The input buffer copying looks good. I've made some notes about the query pool handling though - there's one wrinkle I hadn't thought of which makes it more complex, but hopefully some ways it can be simplified so it's not too bad overall.

baldurk commented on renderdoc/driver/vulkan/wrappers/vk_cmd_funcs.cpp:
```diff
@@ -7873,6 +7920,15 @@ void WrappedVulkan::vkCmdBuildAccelerationStructuresKHR(
+      if(copyResult != ResultCode::Succeeded)
+      {
+        m_LastCaptureError = copyResult;
+        RDCERR("%s", copyResult.message.c_str());
```
This error print is no longer needed with the `RETURN_ERROR_RESULT` macro (which prints it at the source, so better file/line in the log.

baldurk commented on renderdoc/driver/vulkan/wrappers/vk_cmd_funcs.cpp:
```diff
@@ -4546,6 +4546,53 @@ void WrappedVulkan::vkCmdCopyQueryPoolResults(VkCommandBuffer commandBuffer, VkQ
+        rdcarray<VkBufferCopy> regions;
+        regions.reserve(queryCount);
+        for(size_t i = 0; i < queryCount; ++i)
+          regions.push_back(VkBufferCopy{(firstQuery + i) * sizeof(uint64_t),
```
Please don't use this altered form of construction. You should be able to drop it here and just use braces for a POD struct, or else you could make a temporary struct outside the loop and update members inside (or resize + set, if you prefer).

baldurk commented on renderdoc/driver/vulkan/wrappers/vk_cmd_funcs.cpp:
```diff
@@ -4546,6 +4546,53 @@ void WrappedVulkan::vkCmdCopyQueryPoolResults(VkCommandBuffer commandBuffer, VkQ
+                        regions.data());
+      }
+
+      if(flags & VK_QUERY_RESULT_WITH_AVAILABILITY_BIT)
```
Unfortunately the availability bit indicates a _per-query_ availability - so `uint64 size; uint64 avail;` etc. If this flag is passed we'll need to go through the `else` path above to split up copies and interleave with updates.

Again though I would not expect anyone to actually use this as it's mostly for things like occlusion queries that aren't waited on or such, not a size that you really really need.

cmannett85-arm replied:
> Again though I would not expect anyone to actually use this

Yeah but the spec says they can and if someone does they'll read past the end of the results array (because it's half the size they expect) and complain.

baldurk replied:
Definitely, I just mean it's not a problem to put it in the suboptimal bucket with non-64bit results by doing separate copies/updates.

baldurk commented on renderdoc/driver/vulkan/wrappers/vk_cmd_funcs.cpp:
```diff
@@ -4539,13 +4539,63 @@ void WrappedVulkan::vkCmdCopyQueryPoolResults(VkCommandBuffer commandBuffer, VkQ
+      };
+      vkCmdPipelineBarrier(commandBuffer, VK_PIPELINE_STAGE_TRANSFER_BIT,
+                           VK_PIPELINE_STAGE_TOP_OF_PIPE_BIT, 0, 1, &barrier, 0, VK_NULL_HANDLE, 0,
+                           VK_NULL_HANDLE);
```
One other minor thing here - we should call directly into the unwrapped versions of all these commands as necessary. Otherwise these will 'look like' application calls and will be serialised into the command buffers and show up in captures.

baldurk commented on renderdoc/driver/vulkan/vk_resources.cpp:
```diff
@@ -3948,6 +3948,19 @@ void ImgRefs::Split(bool splitAspects, bool splitLevels, bool splitLayers)
+QueryPoolInfo::QueryPoolInfo(WrappedVulkan *driver, VkDevice device,
+                             const VkQueryPoolCreateInfo *pCreateInfo)
+{
+  m_Buffer.Create(driver, device, pCreateInfo->queryCount * 8, 1, GPUBuffer::eGPUBufferReadback);
```
Something I forgot before - this needs a call to `ResourceManager::SetInternalResource` to ensure this buffer is not considered a user buffer, even though it's created and wrapped. Otherwise if someone enables 'ref all resources' this buffer will show up in the capture visibly. See how these buffers are created in the text renderer for an example of how this is handled for our internal buffers during capture.

baldurk commented on renderdoc/driver/vulkan/wrappers/vk_cmd_funcs.cpp:
```diff
@@ -4539,13 +4539,66 @@ void WrappedVulkan::vkCmdCopyQueryPoolResults(VkCommandBuffer commandBuffer, VkQ
+          VK_ACCESS_TRANSFER_READ_BIT,
+      };
+      ObjDisp(commandBuffer)
+          ->CmdPipelineBarrier(commandBuffer, VK_PIPELINE_STAGE_TRANSFER_BIT,
```
These calls now need `Unwrap()`s for the command buffer and buffers, since we're no longer going to do it automatically going through our capture hook.

cmannett85-arm committed:
Switch to GPU side query result replacement
